core/state/state_object.go

```diff
@@ -311,28 +335,49 @@ func (s *StateObject) GetCommittedState(db Database, key common.Hash) common.Has
 		value.SetBytes(content)
 	}
 	s.setOriginStorage(key, value)
-	return value
+	return value, nil
 }
 
 // SetState updates a value in account storage.
-func (s *StateObject) SetState(db Database, key, value common.Hash) {
+func (s *StateObject) SetState(db Database, key, value common.Hash) error {
 	// If the fake storage is set, put the temporary state update here.
 	if s.fakeStorage != nil {
 		s.fakeStorage[key] = value
-		return
+		return nil
 	}
 	// If the new value is the same as old, don't set
-	prev := s.GetState(db, key)
+	prev, err := s.GetState(db, key)
+	if exErr, ok := err.(*ExpiredStateError); ok {
+		exErr.isInsert = true
+		return exErr
+	}
+	if err != nil {
+		return err
+	}
 	if prev == value {
-		return
+		s.accessState(key)
+		return nil
+	}
+	// when state insert, check if valid to insert new state
+	if prev != (common.Hash{}) {
+		_, err = s.getDirtyReviveTrie(db).TryGet(key.Bytes())
+		if err != nil {
+			if enErr, ok := err.(*trie.ExpiredNodeError); ok {
+				return NewInsertExpiredStateError(s.address, enErr)
```

Comment:
I think `NewInsertExpiredStateError` still needs to store the `key` instead of the empty `key`, so that it can be used to generate proof